### ISSUE TITLE
Unable to use `read_json` with nested STRUCT data types yields unexpected end of data

### ISSUE THREAD

**Issue by jaanli** [duckdb/duckdb#7015]
### What happens?

Trying to execute the following: 

```
%%sql 
SELECT *
FROM read_json('./data/datathinking.zulipchat.com/raw/messages-000001.json',
               json_format='array_of_records',
               maximum_object_size=:file_size_in_bytes,
               columns={
                   'zerver_message': 'STRUCT(id BIGINT, 
                                             sender BIGINT, 
                                             recipient BIGINT, 
                                             realm BIGINT, 
                                             subject VARCHAR, 
                                             content VARCHAR, rendered_content VARCHAR, rendered_content_version BIGINT, 
                                             date_sent TIMESTAMP, sending_client BIGINT, last_edit_time TIMESTAMP, edit_history VARCHAR, has_attachment BOOLEAN, has_image BOOLEAN, 
                                             has_link BOOLEAN, search_tsvector VARCHAR)',
                    'zerver_usermessage': 'STRUCT(id BIGINT, 
                                                  user_profile BIGINT, 
                                                  message BIGINT, 
                                                  flags_mask BIGINT)'
                        },
               lines='true'
);
```

In this environment: https://github.com/onefact/datathinking.org-codespace/blob/main/notebooks/in-class-notebooks/230410-data-build-tool-for-standardizing-zulip-chat-data.ipynb 

Leads to this error: 

```
(duckdb.InvalidInputException) Invalid Input Error: Malformed JSON in file "[./data/datathinking.zulipchat.com/raw/messages-000001.json](https://file+.vscode-resource.vscode-cdn.net/Users/me/dropbox/projects/datathinking.org-codespace-main/notebooks/in-class-notebooks/data/datathinking.zulipchat.com/raw/messages-000001.json)", at byte 2 in line 2: unexpected end of data. 
(Background on this error at: https://sqlalche.me/e/14/f405)
```

### To Reproduce

Execute this: 
```
%%sql 
SELECT *
FROM read_json('./data/datathinking.zulipchat.com/raw/messages-000001.json',
               json_format='array_of_records',
               maximum_object_size=:file_size_in_bytes,
               columns={
                   'zerver_message': 'STRUCT(id BIGINT, 
                                             sender BIGINT, 
                                             recipient BIGINT, 
                                             realm BIGINT, 
                                             subject VARCHAR, 
                                             content VARCHAR, rendered_content VARCHAR, rendered_content_version BIGINT, 
                                             date_sent TIMESTAMP, sending_client BIGINT, last_edit_time TIMESTAMP, edit_history VARCHAR, has_attachment BOOLEAN, has_image BOOLEAN, 
                                             has_link BOOLEAN, search_tsvector VARCHAR)',
                    'zerver_usermessage': 'STRUCT(id BIGINT, 
                                                  user_profile BIGINT, 
                                                  message BIGINT, 
                                                  flags_mask BIGINT)'
                        },
               lines='true'
);
```

Against this file: https://raw.githubusercontent.com/onefact/datathinking.org-codespace/main/data/datathinking.zulipchat.com/raw/messages-000001.json

### OS:

Mac

### DuckDB Version:

0.7.1

### DuckDB Client:

Python

### Full Name:

Jaan Altosaar Li

### Affiliation:

University of Tartu & One Fact Foundation

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by jaanli** [comment]
Just following up! We are still blocked in class and trying to ask GPT-4 for help here :(

**Comment by Mytherin** [comment]
This looks like a duplicate of #6569

**Comment by Mause** [comment]
I'm not certain it is, as the JSON they're trying to load has an object at the root, instead of an array?

**Comment by Mytherin** [comment]
Ah yes, you are right. Then the option `array_of_records` is incorrect as it is not an array of records but rather a single large object. This works fine for me out-of-the-box with the auto-detect:

```sql
D SELECT * FROM read_json_auto('messages-000001.json', maximum_object_size=99999999);
┌──────────────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│    zerver_message    │                                                                          zerver_usermessage                                                                           │
│ struct(id ubigint,…  │                                            struct(id ubigint, user_profile ubigint, message ubigint, flags_mask ubigint)[]                                            │
├──────────────────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ [{'id': 299488052,…  │ [{'id': 8257034355, 'user_profile': 544719, 'message': 299488052, 'flags_mask': 1}, {'id': 8257034356, 'user_profile': 544719, 'message': 299488053, 'flags_mask': …  │
└──────────────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

**Comment by Mytherin** [comment]
I would suggest [having a look at the JSON documentation](https://duckdb.org/docs/data/json/overview#parameters). `lines`  should only be set to true for [ndjson](http://ndjson.org). And [array_of_records](https://duckdb.org/docs/data/json/overview#array_of_records) requires a single array at the bottom. The behavior of the system seems correct to me in this case. @lnkuiper perhaps we can make the error message here more clear?

**Comment by jaanli** [comment]
> Ah yes, you are right. Then the option `array_of_records` is incorrect as it is not an array of records but rather a single large object. This works fine for me out-of-the-box with the auto-detect:
> 
> ```sql
> D SELECT * FROM read_json_auto('messages-000001.json', maximum_object_size=99999999);
> ┌──────────────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
> │    zerver_message    │                                                                          zerver_usermessage                                                                           │
> │ struct(id ubigint,…  │                                            struct(id ubigint, user_profile ubigint, message ubigint, flags_mask ubigint)[]                                            │
> ├──────────────────────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
> │ [{'id': 299488052,…  │ [{'id': 8257034355, 'user_profile': 544719, 'message': 299488052, 'flags_mask': 1}, {'id': 8257034356, 'user_profile': 544719, 'message': 299488053, 'flags_mask': …  │
> └──────────────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
> ```

Thank you so much @Mytherin! 

Unfortunately this works but does not seem to understand the `STRUCT` fields that we entered into the `read_json` function after reading the documentation here: https://duckdb.org/2023/03/03/json.html#handling-inconsistent-json-schemas

Is there other documentation besides this blog post on how to read nested json schemas? 

We are still blocked on this unfortunately.

**Comment by jaanli** [comment]
Just following up on this if anyone has any advice!

**Comment by Mytherin** [comment]
There is only one JSON object because your file contains only one JSON object. You are probably looking for [unnest](https://duckdb.org/docs/sql/query_syntax/unnest.html) to extract the elements from the nested arrays.

**Comment by jaanli** [comment]
That worked!!!! Thank you so very much @Mytherin! 

Here is the workaround: 

https://github.com/onefact/datathinking.org-codespace/blob/main/notebooks/in-class-notebooks/230420-zulip-chat-duckdb-loading.ipynb

Hope this helps others while we figure out how to get the original `read_json` to work :D

**Comment by lnkuiper** [comment]
Closed via #7478
